File: src/lua/types/lttorrenthandle.cpp
#include "../types.hpp"

#include <libtorrent/torrent_handle.hpp>

#include "../../torrentclientdata.hpp"

using porla::Lua::Types::LtTorrentHandle;

void LtTorrentHandle::Register(sol::state& lua)
{
    lua.new_usertype<lt::torrent_handle>(
        "lt.torrent_handle",
        sol::no_constructor,
        // add_piece
        // add_tracker
        "add_url_seed",               &lt::torrent_handle::add_url_seed,
        "clear_error",                &lt::torrent_handle::clear_error,
        "clear_peers",                &lt::torrent_handle::clear_peers,
        "clear_piece_deadlines",      &lt::torrent_handle::clear_piece_deadlines,
        // connect_peer
        "download_limit",             &lt::torrent_handle::download_limit,
        // file_priority
        // file_progress
        "file_status",                &lt::torrent_handle::file_status,
        // flags
        "flush_cache",                &lt::torrent_handle::flush_cache,
        "force_dht_announce",         &lt::torrent_handle::force_dht_announce,
        "force_lsd_announce",         &lt::torrent_handle::force_lsd_announce,
        // force_reannounce
        "force_recheck",              &lt::torrent_handle::force_recheck,
        //"get_download_queue",       &lt::torrent_handle::get_download_queue,
        // get_file_priorities
        // get_peer_info
        "get_peer_info",              [](const lt::torrent_handle& th)
                                        {
                                            std::vector<lt::peer_info> peers;
                                            th.get_peer_info(peers);
                                            return peers;
                                        },
        // get_piece_priorities
        "get_renamed_files",          &lt::torrent_handle::get_renamed_files,
        // get_resume_data
        "have_piece",                 &lt::torrent_handle::have_piece,
        "in_session",                 &lt::torrent_handle::in_session,
        "info_hash",                  &lt::torrent_handle::info_hashes,
        "is_valid",                   &lt::torrent_handle::is_valid,
        "max_connections",            &lt::torrent_handle::max_connections,
        "max_uploads",                &lt::torrent_handle::max_uploads,
        "move_storage",               &lt::torrent_handle::move_storage,
        // "need_save_resume_data"
        "pause",                      &lt::torrent_handle::pause,
        // piece_availability
        //"piece_layers",               &lt::torrent_handle::piece_layers,
        // piece_priority
        "post_download_queue",        &lt::torrent_handle::post_download_queue,
        "post_file_priorities",       &lt::torrent_handle::post_file_priorities,
        "post_file_progress",         &lt::torrent_handle::post_file_progress,
        "post_file_status",           &lt::torrent_handle::post_file_status,
        "post_peer_info",             &lt::torrent_handle::post_peer_info,
        "post_piece_availability",    &lt::torrent_handle::post_piece_availability,
        "post_status",                &lt::torrent_handle::post_status,
        "post_trackers",              &lt::torrent_handle::post_trackers,
        // prioritize_files
        // prioritize_pieces
        "queue_position",             &lt::torrent_handle::queue_position,
        "queue_position_bottom",      &lt::torrent_handle::queue_position_bottom,
        "queue_position_down",        &lt::torrent_handle::queue_position_down,
        "queue_position_set",         &lt::torrent_handle::queue_position_set,
        "queue_position_top",         &lt::torrent_handle::queue_position_top,
        "queue_position_up",          &lt::torrent_handle::queue_position_up,
        "read_piece",                 [](const lt::torrent_handle& th, int piece_index)
                                        { th.read_piece(lt::piece_index_t{piece_index}); },
        "remove_url_seed",            &lt::torrent_handle::remove_url_seed,
        "rename_file",                &lt::torrent_handle::rename_file,
        // replace_trackers
        // reset_piece_deadline
        "resume",                     &lt::torrent_handle::resume,
        // save_resume_data
        // scrape_tracker
        "set_download_limit",         &lt::torrent_handle::set_download_limit,
        // set_flags
        "set_max_connections",        &lt::torrent_handle::set_max_connections,
        "set_max_uploads",            &lt::torrent_handle::set_max_uploads,
        // set_metadata
        // set_piece_deadline
        // set_sequential_range
        "set_ssl_certificate",        &lt::torrent_handle::set_ssl_certificate,
        "set_ssl_certificate_buffer", &lt::torrent_handle::set_ssl_certificate_buffer,
        "set_upload_limit",           &lt::torrent_handle::set_upload_limit,
        "status",                     [](const lt::torrent_handle& th) { return th.status(); },
        "torrent_file",               &lt::torrent_handle::torrent_file,
        "trackers",                   &lt::torrent_handle::trackers,
        // unset_flags
        "upload_limit",               &lt::torrent_handle::upload_limit,
        "url_seeds",                  &lt::torrent_handle::url_seeds);
}

File: types/lua/usertypes/lttorrenthandle.lua
---@meta

---@class LtTorrentHandle
local LtTorrentHandle = {}

---@param url string
function LtTorrentHandle:add_url_seed(url) end

function LtTorrentHandle:clear_error() end

function LtTorrentHandle:clear_peers() end

function LtTorrentHandle:clear_piece_deadlines() end

---@return integer
function LtTorrentHandle:download_limit() end

---@return LtOpenFileState[]
function LtTorrentHandle:file_status() end

---@return LtTorrentFlags
function LtTorrentHandle:flags() end

function LtTorrentHandle:flush_cache() end

function LtTorrentHandle:force_dht_announce() end

function LtTorrentHandle:force_lsd_announce() end

---@class LtReannounceFlagsT
---@field ignore_min_interval? boolean
---@field high_priority? boolean

---@param seconds? integer
---@param index? integer
---@param flags? LtReannounceFlagsT
function LtTorrentHandle:force_reannounce(seconds, index, flags) end

function LtTorrentHandle:force_recheck() end

---@return integer[]
function LtTorrentHandle:get_file_priorities() end

---@return LtPeerInfo[]
function LtTorrentHandle:get_peer_info() end

---@return LtRenamedFiles
function LtTorrentHandle:get_renamed_files() end

---@param index integer
---@return boolean
function LtTorrentHandle:have_piece(index) end

---@return boolean
function LtTorrentHandle:in_session() end

---@return LtInfoHash
function LtTorrentHandle:info_hash() end

---@return boolean
function LtTorrentHandle:is_valid() end

---@return integer
function LtTorrentHandle:max_connections() end

---@return integer
function LtTorrentHandle:max_uploads() end

---@alias LtMoveFlags
---| "always_replace_files"
---| "fail_if_exist"
---| "dont_replace"
---| "reset_save_path"
---| "reset_save_path_unchecked"

---@class LtMoveStorageOpts
---@field flags? LtMoveFlags

---@param path string
---@param opts? LtMoveStorageOpts
function LtTorrentHandle:move_storage(path, opts) end

function LtTorrentHandle:pause() end

function LtTorrentHandle:post_download_queue() end

function LtTorrentHandle:post_file_priorities() end

function LtTorrentHandle:post_file_progress() end

function LtTorrentHandle:post_file_status() end

function LtTorrentHandle:post_peer_info() end

function LtTorrentHandle:post_piece_availability() end

function LtTorrentHandle:post_status() end

function LtTorrentHandle:post_trackers() end

---@param prios integer[]
function LtTorrentHandle:prioritize_files(prios) end

function LtTorrentHandle:queue_position() end

function LtTorrentHandle:queue_position_bottom() end

function LtTorrentHandle:queue_position_down() end

---@param pos integer
function LtTorrentHandle:queue_position_set(pos) end

function LtTorrentHandle:queue_position_top() end

function LtTorrentHandle:queue_position_up() end

---@param index integer
function LtTorrentHandle:read_piece(index) end

---@param url string
function LtTorrentHandle:remove_url_seed(url) end

---@param index integer
---@param name string
function LtTorrentHandle:rename_file(index, name) end

function LtTorrentHandle:resume() end

function LtTorrentHandle:save_resume_data() end

---@param limit integer
function LtTorrentHandle:set_download_limit(limit) end

---@param flags LtTorrentFlags
---@param mask? LtTorrentFlags
function LtTorrentHandle:set_flags(flags, mask) end

---@param n integer
function LtTorrentHandle:set_max_connections(n) end

---@param n integer
function LtTorrentHandle:set_max_uploads(n) end

---@param certificate string
---@param private_key string
---@param dh_params string
---@param passphrase string
function LtTorrentHandle:set_ssl_certificate(certificate, private_key, dh_params, passphrase) end

---@param buffer string
---@param private_key string
---@param dh_params string
---@param passphrase string
function LtTorrentHandle:set_ssl_certificate_buffer(buffer, private_key, dh_params, passphrase) end

---@param limit integer
function LtTorrentHandle:set_upload_limit(limit) end

---@return LtTorrentStatus
function LtTorrentHandle:status() end

---@return LtTorrentInfo
function LtTorrentHandle:torrent_file() end

---@return LtAnnounceEntry[]
function LtTorrentHandle:trackers() end

---@return integer
function LtTorrentHandle:upload_limit() end

---@return string[]
function LtTorrentHandle:url_seeds() end

---@return PoTorrentData
function LtTorrentHandle:userdata() end
